### DISCUSSION TITLE
Log formatter and handler

### DISCUSSION THREAD

**Discussion by vfresnaud** [pymeasure/pymeasure#886] [Q&A]
Hello,
is there a way to specify the logger, or at least the formatter? (console.log)
I know that it is described in log.py, but it will be nice to have a setter to change this formatter and to use our own (for instance, we prefer to display the LEVEL of log at the beginning of the message, not at the end. It allows to see the LEVEL always at the same indentation).
BR,
Vincent

Example below:
```
formatter = '%(asctime)s - %(name)s - %(levelname)s - [%(message)s] - (%(filename)s:%(lineno)d)'

and adding colors to the messages:

class CustomFormatter(logging.Formatter): #pragma:no cover

    cyan_foreground = '\x1b[36m'
    violet_foreground = '\x1b[7;49;95m'
    blue_foreground = '\x1b[38;5;39m'
    yellow_foreground = '\x1b[38;5;226m'
    red_foreground = '\x1b[38;5;196m'
    bold_red_foreground = '\x1b[31;1m'
    reset_foreground = '\x1b[0m'

    def __init__(self, fmt):
        super().__init__()
        self.fmt = fmt
        self.FORMATS = {
            logging.DEBUG: self.cyan_foreground + self.fmt + self.reset_foreground,
            logging.INFO: self.blue_foreground + self.fmt + self.reset_foreground,
            logging.WARNING: self.yellow_foreground + self.fmt + self.reset_foreground,
            logging.ERROR: self.red_foreground + self.fmt + self.reset_foreground,
            logging.CRITICAL: self.bold_red_foreground + self.fmt + self.reset_foreground
        }

    def format(self, record):
        log_fmt = self.FORMATS.get(record.levelno)
        formatter = logging.Formatter(log_fmt)
        return formatter.format(record)
```

**Comment by CasperSchippers** [comment]
That is a very good question. I don't think there is a ready-made option for this.
You can change the formatter of the logger in the `init` method of the `ManagedWindow` class in your script (I used the basic gui example below).

```Python
class MainWindow(ManagedWindow):

    def __init__(self):
        super().__init__(
            procedure_class=TestProcedure,
            inputs=['iterations', 'delay', 'seed'],
            displays=['iterations', 'delay', 'seed'],
            x_axis='Iteration',
            y_axis='Random Number'
        )
        self.setWindowTitle('GUI Example')

        # Change the formatter to a new one, or a custom one
        self.log_widget.handler.setFormatter(logging.Formatter(formatter))
```
There is already 1 logging message emitted though....

One solution would be to allow passing kwargs to the log_widget and use that to set the formatter upon initialization, but that is not possible at this moment. Please also let us know if you have another solution, and feel free to open a PR with a solution if you want to.

Regarding the colouring that you do in your example: For a 'recent' project, I also had to do something like this; I was planning to make PR out of it, but didn't yet get around to do that, so i'll try doing that asap.

**Reply by CasperSchippers** [reply]
I quickly made a PR #890, which could help (at least adds the colouring and the passing of the format-string); see what you think of that.

**Reply by vfresnaud** [reply]
Hello,
 yes that looks great!
We will definitively try it very soon, i will let you know.
Very good idea the blinking tab!
thank you for this!
